discovery/README.md

```diff
@@ -0,0 +1,123 @@
+that a machine is going to be a Kafka server, likely a machine database or
+configuration management system.
+
+If something is particlarly custom or unusual, `file_sd` is the generic
+mechanism provided for users to hook in. Generally with Prometheus we offer a
+single generic mechanism for things with infinite variations, rather than
+trying to support everything natively (see also, alertmanager webhook, remote
+read, remote write, node exporter textfile collector). For example anything
+that would involve talking to a relational database should use `file_sd`
+instead.
+
+For configuration management systems like Chef, while they do have a
+database/API that'd in principle make sense to talk to for service discovery,
+the idomatic approach is to use Chef's templating facilities to write out a
+file for use with `file_sd`.
+
+
+### Mapping from SD to Prometheus
+
+The general principle with SD is to extract all the potentially useful
+information we can out of the SD, and let the user choose what they need of it
+using relabelling. This information is generally termed metadata.
+
+Metadata is exposed as a set of key/value pairs (labels) per target. The keys
+are prefixed with `__meta_sdname_`, and there should also be an `__address__`
+label with the host:port of the target (preferably an IP address to avoid DNS
+lookups). No other labelnames should be exposed.
+
+It is very common for initial pull requests for new SDs to include hardcoded
+assumptions that make sense for the the author's setup. SD should be generic,
+any customisation should be handled via relabelling. There should be basically
+no business logic, filtering, or transformations of the data from the SD beyond
+that which is needed to fit it into the metadata data model. 
+
+Arrays (e.g. a list of tags) should be converted to a single label with the
+array values joined with a comma. Also prefix and suffix the value with a
+comma. So for example the array `[a, b, c]` would become `,a,b,c,`. As
+relabelling regexes are fully anchored, this makes it easier to write correct
+regexes against (`.*,a,.*` works no matter where `a` appears in the list). The
+canonical example of this is `__meta_consul_tags`.
+
+Maps, hashes and other forms of key/value pairs should be all prefixed and
+exposed as labels. For example for EC2 tags, there would be
+`__meta_ec2_tag_Description=mydescription` for the Description tag. Labelnames
+may only contain `[_a-zA-Z0-9]`, sanitize by replacing with underscores as needed.
+
+For targets with multiple potential ports, you can a) expose them as a list, b)
+if they're named expose them as a map or c) expose them each as their own
+target. Kuberenetes SD takes the target per port approach. a) and b) can be
+combined.
+
+For machine-like SDs (OpenStack, EC2, Kubernetes to some extent) there may
+be multiple network interfaces for a target. Thus far reporting the details
+of only the first network interface has sufficed.
```

Comment:
"First" is a bit fuzzy. For Kubernetes for example there is no "first" but Kubernetes itself defines a precedence order elsewhere which we copied over.